django/http/request.py

```diff
@@ -350,49 +353,52 @@ def _mark_post_parse_error(self):
         self._post = QueryDict()
         self._files = MultiValueDict()
 
-    def _load_post_and_files(self):
-        """Populate self._post and self._files if the content-type is a form type"""
-        if self.method != "POST":
+    def _load_post_and_files(
+        self, data_attr="_post", parsers=POST_PARSERS, methods=("POST",)
+    ):
+        if methods and self.method not in methods:
             self._post, self._files = (
                 QueryDict(encoding=self._encoding),
                 MultiValueDict(),
             )
             return
+
         if self._read_started and not hasattr(self, "_body"):
-            self._mark_post_parse_error()
+            setattr(self, data_attr, QueryDict())
+            self._files = MultiValueDict()
             return
 
-        if self.content_type == "multipart/form-data":
-            if hasattr(self, "_body"):
-                # Use already read data
-                data = BytesIO(self._body)
-            else:
-                data = self
+        selected_parser = None
+        for parser in parsers:
+            if parser.can_handle(self.content_type):
+                selected_parser = parser
+                break
+
+        if selected_parser:
+            selected_parser.parsers = parsers
             try:
-                self._post, self._files = self.parse_file_upload(self.META, data)
-            except (MultiPartParserError, TooManyFilesSent):
+                if selected_parser._supports_form_parsing:
+                    # TODO Not sure how to make these consistent.
+                    data, self._files = parser.parse(None, self)
+                    setattr(self, data_attr, data)
+                else:
+                    data, self._files = parser.parse(self.body, self)
+                    setattr(self, data_attr, data)
```

Comment:
This is what I'm currently focusing on / struggling with / welcome thoughts on. 

Ideally the `parse` method would have the signature `parse(request.body)`. But the form parsers also need to know about the request, so maybe `parse(request.body, request)`. That's a little repetitive so the previous version I had just `parse(request)`. Which is fine, that is until you want to parse just part of a multipart request, then you really just want to pass the content that needs parsing.

Unfortunately passing `self.body` to `parse()` causes the body to be read too early, and that causes test failures. I'd like to remove this little dance here I've created. 🕺

Reply:
hmm... Maybe doing the check in `parse()` is the way to go...

``` python 
class JSONParser(BaseParser):
    media_type = "application/json"

    def parse(self, request):
        from django.http import HttpRequest

        if isinstance(request, HttpRequest):
            request = request.body
        # TODO enable strict mode. Like DRF.
        return json.loads(request), MultiValueDict()
```

Reply:
> But the form parsers also need to know about the request...

Parser instances are per-request right? So, can the parsers take what information they need from the request when being instantiated, thereby not requiring the request (again) at the point of parsing. 

e.g. `FormParser` requires the request encoding, but would take that early (still deferring the check until it's actually called on to try and parse the body). 

🤔

Reply:
Yes, makes sense. 

Will dig more into the test failures from early accessing of `request.body`. 🕵‍♂️

Reply:
So... If a `POST` request has a file and it's larger than [`settings.DATA_UPLOAD_MAX_MEMORY_SIZE`](https://docs.djangoproject.com/en/5.0/ref/settings/#data-upload-max-memory-size) (2.5mb), on accessing `request.body` you'll get:

```
*** django.core.exceptions.RequestDataTooBig: Request body exceeded settings.DATA_UPLOAD_MAX_MEMORY_SIZE.
```

Of course `MultiPartParser` knows all about this and therefore you can upload files bigger than `DATA_UPLOAD_MAX_MEMORY_SIZE`. 

In the most recent commit I have allowed the parsers to accept `request` at `__init__()` and `parse()` becomes `parse(data)`. There's a check added for the content type of `"multipart/form-data"` to avoid accessing `request.body` so that files can be handled correctly. 

Hope that's a step in the right direction. Appreciate all thoughts and comments.